inlang/source-code/sdk/src/createMessageLintReportsQuery.ts

```diff
@@ -1,50 +1,78 @@
 import { ReactiveMap } from "./reactivity/map.js"
+import { debounce } from "throttle-debounce"
+import { createEffect } from "./reactivity/solid.js"
 
 /**
  * Creates a reactive query API for messages.
  */
 export function createMessageLintReportsQuery(
-	messages: () => Array<Message> | undefined,
+	messagesQuery: MessageQueryApi,
 	settings: () => ProjectSettings,
 	installedMessageLintRules: () => Array<InstalledMessageLintRule>,
-	resolvedModules: () => Awaited<ReturnType<typeof resolveModules>> | undefined
+	resolvedModules: () => Awaited<ReturnType<typeof resolveModules>> | undefined,
+	hasWatcher: boolean
 ): InlangProject["query"]["messageLintReports"] {
 	// @ts-expect-error
 	const index = new ReactiveMap<MessageLintReport["messageId"], MessageLintReport[]>()
 
-	createEffect(() => {
-		const modules = resolvedModules()
-		const _messages = messages()
-		const _settings = settings()
+	const modules = resolvedModules()
 
-		if (_messages && _settings && modules) {
-			// index.clear()
-			for (const message of _messages) {
-				// TODO: only lint changed messages and update arrays selectively
+	const rulesArray = modules?.messageLintRules
+	const messageLintRuleLevels = Object.fromEntries(
+		installedMessageLintRules().map((rule) => [rule.id, rule.level])
+	)
+	const settingsObject = () => {
+		return {
+			...settings(),
+			messageLintRuleLevels,
+		}
+	}
+
+	const messages = messagesQuery.getAll() as Message[]
 
-				lintSingleMessage({
-					rules: modules.messageLintRules,
-					settings: {
-						..._settings,
-						messageLintRuleLevels: Object.fromEntries(
-							installedMessageLintRules().map((rule) => [rule.id, rule.level])
-						),
-					},
-					messages: _messages,
-					message: message,
-				}).then((report) => {
-					if (
-						report.errors.length === 0 &&
-						JSON.stringify(index.get(message.id)) !== JSON.stringify(report.data)
-					) {
-						index.set(message.id, report.data || [])
+	createEffect(() => {
+		if (rulesArray) {
+			for (const messageId of messagesQuery.includedMessageIds()) {
+				messagesQuery.get.subscribe({ where: { id: messageId } }, async (message) => {
+					if (hasWatcher) {
+						await lintSingleMessage({
+							rules: rulesArray,
+							settings: settingsObject(),
+							messages: messages,
+							message: message,
+						}).then((report) => {
+							if (report.errors.length === 0 && index.get(messageId) !== report.data) {
+								index.set(messageId, report.data)
+							}
+						})
+					} else {
+						debounce(
+							500,
+							async (message) => {
+								await lintSingleMessage({
+									rules: rulesArray,
+									settings: settingsObject(),
+									messages: messages,
+									message: message,
+								}).then((report) => {
+									if (report.errors.length === 0 && index.get(messageId) !== report.data) {
+										index.set(messageId, report.data)
+									}
+								})
+							},
+							{ atBegin: false }
+						)(message)
```

Comment:
1. Why a debouncer here?
2. Async then problem as above.

Reply:
1. Debouncing is required until this is implemented https://github.com/inlang/monorepo/issues/1817
**Background:** Linting is triggered on every key change; no longer on the debounced save. This works for projects with only a few languages and lint rules, but for projects with 20+ languages and 4 lint rules, it blocks the main thread and the pattern editor field lags massively behind (loss of real-time feedback).

2. Agreed, I'll remove it